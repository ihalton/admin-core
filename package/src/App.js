import React from "react";
import ReactDOM from "react-dom";
import {
  Box,
  Drawer,
  Container,
  List,
  ListItem,
  ListItemText,
  Typography,
  makeStyles
} from "@material-ui/core";
import { Route, Switch, Link, useHistory } from "react-router-dom";
import useAuth from "./hooks/useAuth";

const activeClassName = "active";

const useStyles = makeStyles((theme) => ({
  closeButton: {
    "color": theme.palette.colors.white,
    "backgroundColor": theme.palette.colors.darkBlue500,
    "&:hover": {
      "backgroundColor": theme.palette.colors.darkBlue600,
      // Reset on touch devices, it doesn't add specificity
      "@media (hover: none)": {
        backgroundColor: theme.palette.colors.darkBlue500
      }
    }
  },
  icon: {
    minWidth: 32,
    display: "flex",
    justifyContent: "center",
    marginRight: theme.spacing(2),
    color: theme.palette.colors.coolGrey300
  },
  iconActive: {
    color: theme.palette.text.active
  },
  shopLogo: {
    flex: 1,
    marginRight: theme.spacing(2)
  },
  toolbar: {
    paddingLeft: theme.spacing(2),
    paddingRight: theme.spacing(2)
  },
  listItem: {
    "paddingLeft": theme.spacing(2),
    "paddingRight": theme.spacing(2),
    "&:hover": {
      backgroundColor: theme.palette.colors.darkBlue600,
      transition: `background-color ${theme.transitions.duration.shortest} ${theme.transitions.easing.easeInOut}`
    }
  },
  listItemText: {
    paddingLeft: 0,
    fontSize: theme.typography.fontSize,
    lineHeight: 1.5,
    letterSpacing: 0.5,
    color: theme.palette.colors.black15
  },
  listItemNested: {
    "paddingTop": 0,
    "paddingBottom": 0,
    "paddingLeft": theme.spacing(8),
    "&:hover": {
      backgroundColor: theme.palette.colors.darkBlue600,
      transition: `background-color ${theme.transitions.duration.shortest} ${theme.transitions.easing.easeInOut}`
    }
  },
  link: {
    [`&.${activeClassName} span`]: {
      color: theme.palette.text.secondaryActive,
      fontWeight: theme.typography.fontWeightSemiBold
    },
    [`&.${activeClassName} $icon`]: {
      color: theme.palette.text.active
    }
  }
}));

export default function App({ title, plugins }) {
  const history = useHistory();
  const classes = useStyles();
  const auth = useAuth();

  return (
    <Container>
      <Drawer
        open={true}
        variant="persistent"
      >
        <List>
          {plugins.map(({ route, navTitle }, index) => (
            <ListItem
              key={index}
              button
              onClick={() => history.push(route)}
              className={classes.listItem}
            >
              <ListItemText
                disableTypography
                primary={navTitle}
                className={classes.listItemText}
              />
            </ListItem>
          ))}
          <ListItem
            button
            onClick={() => auth.logout()}
            className={classes.listItem}
          >
            <ListItemText
              disableTypography
              primary="Sign Out"
              className={classes.listItemText}
            />
          </ListItem>
        </List>
      </Drawer>
      <div style={{paddingLeft: 320}}>
        <Typography variant="h1">{title || "Reaction Admin"}</Typography>
        <Switch>
          {plugins.map(({ MainComponent, route }, index) => (
            <Route exact path={route} component={MainComponent} key={index} />
          ))}
        </Switch>
      </div>
    </Container>
  )
};
