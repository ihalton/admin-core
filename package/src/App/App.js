import React, { Suspense } from "react";
import PropTypes from "prop-types";
import { ThemeProvider } from "@material-ui/core";
import { defaultTheme } from "@reactioncommerce/catalyst";
import { BrowserRouter } from "react-router-dom";
import { AuthenticationProvider, OidcSecure } from "@axa-fr/react-oidc-context";
import { ApolloProvider } from "react-apollo";
import Dashboard from "../Dashboard";

/**
 * BAse application component containing providers for theme, auth, routing and more
 * @param {Object} props Props for configuring various providers and base components
 * @returns {React.ReactElement} App
 */
function App(props) {
  const {
    DashboardComponent: DashboardComponentProp,
    apolloClient,
    authenticationProviderProps,
    dashboardComponentProps,
    plugins
  } = props;

  const DashboardComponent = DashboardComponentProp || Dashboard;

  return (
    <Suspense fallback={null}>
      <ApolloProvider client={apolloClient}>
        <BrowserRouter>
          <AuthenticationProvider {...authenticationProviderProps}>
            <ThemeProvider theme={defaultTheme}>
              <OidcSecure>
                <DashboardComponent
                  {...dashboardComponentProps}
                  plugins={plugins}
                />
              </OidcSecure>
            </ThemeProvider>
          </AuthenticationProvider>
        </BrowserRouter>
      </ApolloProvider>
    </Suspense>
  );
}

App.propTypes = {
  DashboardComponent: PropTypes.elementType,
  apolloClient: PropTypes.object,
  authenticationProviderProps: PropTypes.object,
  dashboardComponentProps: PropTypes.object,
  plugins: PropTypes.arrayOf(PropTypes.object)
};

export default App;
