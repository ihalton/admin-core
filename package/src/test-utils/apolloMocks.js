import getViewer from "../graphql/queries/viewer.graphql";

export default [
  {
    request: {
      query: getViewer
    },
    result: {
      data: {
        viewer: {
          _id: "1234"
        }
      }
    }
  }
];
