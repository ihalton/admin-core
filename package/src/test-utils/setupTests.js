// // Add async/await features
// import "regenerator-runtime/runtime";

// // add some helpful assertions
import "@testing-library/jest-dom/extend-expect";

// // Polyfill fetch
// import "unfetch/polyfill";
