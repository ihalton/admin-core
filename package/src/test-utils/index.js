
import React from "react";
import PropTypes from "prop-types";
import { render } from "@testing-library/react";
import { ThemeProvider } from "@material-ui/core";
import { MemoryRouter } from "react-router-dom";
import { MockedProvider } from "@apollo/react-testing";
import { defaultTheme } from "@reactioncommerce/catalyst";
import i18n from "i18next";
import { initReactI18next } from "react-i18next";
import apolloMocks from "./apolloMocks";

i18n
  .use(initReactI18next)
  .init({
    resources: {
      en: {
        translation: {}
      }
    },
    lng: "en",
    fallbackLng: "en",

    interpolation: {
      escapeValue: false
    }
  });

/**
 * Component that wraps components with mock providers during testing.
 * @return {Component} - Component wrapped with mock providers
 */
const TestProviders = ({ children }) => (
  <MockedProvider mocks={apolloMocks}>
    <ThemeProvider theme={defaultTheme}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </ThemeProvider>
  </MockedProvider>
);

TestProviders.propTypes = {
  /** React Component */
  children: PropTypes.element.isRequired
};

/**
 * Custom test renderer that wraps all components with the appropriate mock providers.
 * @param {Component} component - React component to render.
 * @param {Object} options - Options.
 * @return {Object} - @see {@link https://testing-library.com/docs/react-testing-library/api#render-result|react-testing-library}
 */
const renderWithProviders = (component, options) => (
  render(component, { wrapper: TestProviders, ...options })
);

export * from "@testing-library/react";
export { renderWithProviders as render };
