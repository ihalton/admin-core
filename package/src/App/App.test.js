/* eslint-disable react/prop-types */
/* eslint-disable react/no-multi-comp */
import React from "react";
import { Route, Switch } from "react-router-dom";
import { render } from "../test-utils";
import App from "./App";

jest.mock("@axa-fr/react-oidc-context", () => ({
  AuthenticationProvider: ({ children }) => children,
  OidcSecure: ({ children }) => children
}));

const mockPlugins = [
  {
    route: "/test-1",
    navTitle: "Test 1",
    // eslint-disable-next-line react/display-name
    MainComponent: () => (<span>Route Test 1</span>)
  }
];

const MockDashboardComponent = ({ title, plugins }) => (
  <div>
    <h1>{title}</h1>
    <Switch>
      {plugins.map(({ MainComponent, route }, index) => (
        <Route exact path={route} component={MainComponent} key={index} />
      ))}
    </Switch>
  </div>
);

test("should be true", () => {
  const { asFragment } = render((
    <App
      DashboardComponent={MockDashboardComponent}
      dashboardComponentProps={{
        title: "Test App"
      }}
      plugins={mockPlugins}
    />
  ));
  expect(asFragment()).toMatchSnapshot();
});
