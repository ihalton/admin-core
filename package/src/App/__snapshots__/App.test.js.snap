// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be true 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Test App
    </h1>
  </div>
</DocumentFragment>
`;
